{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b41a320b",
   "metadata": {},
   "source": [
    "This notebook is primarily to make sure your environment installed properly. If anything in here does not run, certainly nothing else will."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2a15e7d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Global seed set to 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.9.1 anndata==0.8.0 umap==0.5.3 numpy==1.22.4 scipy==1.9.1 pandas==1.5.0 scikit-learn==1.1.2 statsmodels==0.13.2 pynndescent==0.5.7\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "import scvi\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sc.set_figure_params(dpi=100, frameon=False, color_map='Reds', facecolor=None)\n",
    "sc.logging.print_header()\n",
    "assert(scvi.__version__=='0.16.3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "89aed8ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = '/home/icb/yuge.ji/projects/feature-attribution-sc'  # should be changed to shared dir when I can find one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "005cbb82-0327-465c-a92e-38b59f2c2931",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class SCGEN in module scgen._scgen:\n",
      "\n",
      "class SCGEN(scvi.model.base._vaemixin.VAEMixin, scvi.model.base._training_mixin.UnsupervisedTrainingMixin, scvi.model.base._base_model.BaseModelClass)\n",
      " |  SCGEN(adata: anndata._core.anndata.AnnData, n_hidden: int = 800, n_latent: int = 100, n_layers: int = 2, dropout_rate: float = 0.2, **model_kwargs)\n",
      " |  \n",
      " |  Implementation of scGen model for batch removal and perturbation prediction.\n",
      " |  \n",
      " |  Parameters\n",
      " |  ----------\n",
      " |  adata\n",
      " |      AnnData object that has been registered via :meth:`~scgen.SCGEN.setup_anndata`.\n",
      " |  n_hidden\n",
      " |      Number of nodes per hidden layer.\n",
      " |  n_latent\n",
      " |      Dimensionality of the latent space.\n",
      " |  n_layers\n",
      " |      Number of hidden layers used for encoder and decoder NNs.\n",
      " |  dropout_rate\n",
      " |      Dropout rate for neural networks.\n",
      " |  **model_kwargs\n",
      " |      Keyword args for :class:`~scgen.SCGENVAE`\n",
      " |  \n",
      " |  Examples\n",
      " |  --------\n",
      " |  >>> vae = scgen.SCGEN(adata)\n",
      " |  >>> vae.train()\n",
      " |  >>> adata.obsm[\"X_scgen\"] = vae.get_latent_representation()\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      SCGEN\n",
      " |      scvi.model.base._vaemixin.VAEMixin\n",
      " |      scvi.model.base._training_mixin.UnsupervisedTrainingMixin\n",
      " |      scvi.model.base._base_model.BaseModelClass\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, adata: anndata._core.anndata.AnnData, n_hidden: int = 800, n_latent: int = 100, n_layers: int = 2, dropout_rate: float = 0.2, **model_kwargs)\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  batch_removal(self, adata: Optional[anndata._core.anndata.AnnData] = None) -> anndata._core.anndata.AnnData\n",
      " |      Removes batch effects.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      adata\n",
      " |          AnnData object with equivalent structure to initial AnnData. If `None`, defaults to the\n",
      " |          AnnData object used to initialize the model. Must have been setup with `batch_key` and `labels_key`,\n",
      " |          corresponding to batch and cell type metadata, respectively.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      corrected: `~anndata.AnnData`\n",
      " |          AnnData of corrected gene expression in adata.X and corrected latent space in adata.obsm[\"latent\"].\n",
      " |          A reference to the original AnnData is in `corrected.raw` if the input adata had no `raw` attribute.\n",
      " |  \n",
      " |  binary_classifier(self, adata, delta, ctrl_key, stim_key, path_to_save, save=True, fontsize=14)\n",
      " |      Latent space classifier.\n",
      " |      \n",
      " |      Builds a linear classifier based on the dot product between\n",
      " |      the difference vector and the latent representation of each\n",
      " |      cell and plots the dot product results between delta and latent\n",
      " |      representation.\n",
      " |      \n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      adata: `~anndata.AnnData`\n",
      " |          AnnData object with equivalent structure to initial AnnData. If `None`, defaults to the\n",
      " |          AnnData object used to initialize the model. Must have been setup with `batch_key` and `labels_key`,\n",
      " |          corresponding to batch and cell type metadata, respectively.\n",
      " |      delta: float\n",
      " |          Difference between stimulated and control cells in latent space\n",
      " |      ctrl_key: basestring\n",
      " |          key for `control` part of the `data` found in `condition_key`.\n",
      " |      stim_key: basestring\n",
      " |          key for `stimulated` part of the `data` found in `condition_key`.\n",
      " |      path_to_save: basestring\n",
      " |          path to save the plot.\n",
      " |      save: boolean\n",
      " |          Specify if the plot should be saved or not.\n",
      " |      fontsize: integer\n",
      " |          Set the font size of the plot.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> import anndata\n",
      " |      >>> import scgen\n",
      " |      >>> import scanpy as sc\n",
      " |      >>> train = sc.read(\"./tests/data/train.h5ad\", backup_url=\"https://goo.gl/33HtVh\")\n",
      " |      >>> scgen.SCGEN.setup_anndata(train)\n",
      " |      >>> network = scgen.SCGEN(train)\n",
      " |      >>> network.train()\n",
      " |      >>> unperturbed_data = train[((train.obs[\"cell_type\"] == \"CD4T\") & (train.obs[\"condition\"] == \"control\"))]\n",
      " |      >>> pred, delta = network.predict(\n",
      " |      >>>     adata=train,\n",
      " |      >>>     adata_to_predict=unperturbed_data,\n",
      " |      >>>     ctrl_key=\"control\",\n",
      " |      >>>     stim_key=\"stimulated\"\n",
      " |      >>>)\n",
      " |      >>> network.binary_classifier(\n",
      " |      >>>     network,\n",
      " |      >>>     train,\n",
      " |      >>>     delta,\n",
      " |      >>>     ctrl_key=\"control\",\n",
      " |      >>>     stim_key=\"stimulated\",\n",
      " |      >>>     path_to_save=\"tests/binary_classifier.pdf\"\n",
      " |      >>>     )\n",
      " |  \n",
      " |  get_decoded_expression(self, adata: Optional[anndata._core.anndata.AnnData] = None, indices: Optional[Sequence[int]] = None, batch_size: Optional[int] = None) -> numpy.ndarray\n",
      " |      Get decoded expression.\n",
      " |  \n",
      " |  predict(self, ctrl_key=None, stim_key=None, adata_to_predict=None, celltype_to_predict=None, restrict_arithmetic_to='all') -> anndata._core.anndata.AnnData\n",
      " |      Predicts the cell type provided by the user in stimulated condition.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      ctrl_key: basestring\n",
      " |          key for `control` part of the `data` found in `condition_key`.\n",
      " |      stim_key: basestring\n",
      " |          key for `stimulated` part of the `data` found in `condition_key`.\n",
      " |      adata_to_predict: `~anndata.AnnData`\n",
      " |          Adata for unperturbed cells you want to be predicted.\n",
      " |      celltype_to_predict: basestring\n",
      " |          The cell type you want to be predicted.\n",
      " |      restrict_arithmetic_to: basestring or dict\n",
      " |          Dictionary of celltypes you want to be observed for prediction.\n",
      " |      Returns\n",
      " |      -------\n",
      " |      predicted_cells: np nd-array\n",
      " |          `np nd-array` of predicted cells in primary space.\n",
      " |      delta: float\n",
      " |          Difference between stimulated and control cells in latent space\n",
      " |  \n",
      " |  reg_mean_plot(self, adata, axis_keys, labels, path_to_save='./reg_mean.pdf', save=True, gene_list=None, show=False, top_100_genes=None, verbose=False, legend=True, title=None, x_coeff=0.3, y_coeff=0.8, fontsize=14, **kwargs)\n",
      " |      Plots mean matching figure for a set of specific genes.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      adata: `~anndata.AnnData`\n",
      " |          AnnData object with equivalent structure to initial AnnData. If `None`, defaults to the\n",
      " |          AnnData object used to initialize the model. Must have been setup with `batch_key` and `labels_key`,\n",
      " |          corresponding to batch and cell type metadata, respectively.\n",
      " |      axis_keys: dict\n",
      " |          Dictionary of `adata.obs` keys that are used by the axes of the plot. Has to be in the following form:\n",
      " |           `{\"x\": \"Key for x-axis\", \"y\": \"Key for y-axis\"}`.\n",
      " |      labels: dict\n",
      " |          Dictionary of axes labels of the form `{\"x\": \"x-axis-name\", \"y\": \"y-axis name\"}`.\n",
      " |      path_to_save: basestring\n",
      " |          path to save the plot.\n",
      " |      save: boolean\n",
      " |          Specify if the plot should be saved or not.\n",
      " |      gene_list: list\n",
      " |          list of gene names to be plotted.\n",
      " |      show: bool\n",
      " |          if `True`: will show to the plot after saving it.\n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> import anndata\n",
      " |      >>> import scgen\n",
      " |      >>> import scanpy as sc\n",
      " |      >>> train = sc.read(\"./tests/data/train.h5ad\", backup_url=\"https://goo.gl/33HtVh\")\n",
      " |      >>> scgen.SCGEN.setup_anndata(train)\n",
      " |      >>> network = scgen.SCGEN(train)\n",
      " |      >>> network.train()\n",
      " |      >>> unperturbed_data = train[((train.obs[\"cell_type\"] == \"CD4T\") & (train.obs[\"condition\"] == \"control\"))]\n",
      " |      >>> pred, delta = network.predict(\n",
      " |      >>>     adata=train,\n",
      " |      >>>     adata_to_predict=unperturbed_data,\n",
      " |      >>>     ctrl_key=\"control\",\n",
      " |      >>>     stim_key=\"stimulated\"\n",
      " |      >>>)\n",
      " |      >>> pred_adata = anndata.AnnData(\n",
      " |      >>>     pred,\n",
      " |      >>>     obs={\"condition\": [\"pred\"] * len(pred)},\n",
      " |      >>>     var={\"var_names\": train.var_names},\n",
      " |      >>>)\n",
      " |      >>> CD4T = train[train.obs[\"cell_type\"] == \"CD4T\"]\n",
      " |      >>> all_adata = CD4T.concatenate(pred_adata)\n",
      " |      >>> network.reg_mean_plot(\n",
      " |      >>>     all_adata,\n",
      " |      >>>     axis_keys={\"x\": \"control\", \"y\": \"pred\", \"y1\": \"stimulated\"},\n",
      " |      >>>     gene_list=[\"ISG15\", \"CD3D\"],\n",
      " |      >>>     path_to_save=\"tests/reg_mean.pdf\",\n",
      " |      >>>     show=False\n",
      " |      >>> )\n",
      " |  \n",
      " |  reg_var_plot(self, adata, axis_keys, labels, path_to_save='./reg_var.pdf', save=True, gene_list=None, top_100_genes=None, show=False, legend=True, title=None, verbose=False, x_coeff=0.3, y_coeff=0.8, fontsize=14, **kwargs)\n",
      " |      Plots variance matching figure for a set of specific genes.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      adata: `~anndata.AnnData`\n",
      " |          AnnData object with equivalent structure to initial AnnData. If `None`, defaults to the\n",
      " |          AnnData object used to initialize the model. Must have been setup with `batch_key` and `labels_key`,\n",
      " |          corresponding to batch and cell type metadata, respectively.\n",
      " |      axis_keys: dict\n",
      " |          Dictionary of `adata.obs` keys that are used by the axes of the plot. Has to be in the following form:\n",
      " |           `{\"x\": \"Key for x-axis\", \"y\": \"Key for y-axis\"}`.\n",
      " |      labels: dict\n",
      " |          Dictionary of axes labels of the form `{\"x\": \"x-axis-name\", \"y\": \"y-axis name\"}`.\n",
      " |      path_to_save: basestring\n",
      " |          path to save the plot.\n",
      " |      save: boolean\n",
      " |          Specify if the plot should be saved or not.\n",
      " |      gene_list: list\n",
      " |          list of gene names to be plotted.\n",
      " |      show: bool\n",
      " |          if `True`: will show to the plot after saving it.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> import anndata\n",
      " |      >>> import scgen\n",
      " |      >>> import scanpy as sc\n",
      " |      >>> train = sc.read(\"./tests/data/train.h5ad\", backup_url=\"https://goo.gl/33HtVh\")\n",
      " |      >>> scgen.SCGEN.setup_anndata(train)\n",
      " |      >>> network = scgen.SCGEN(train)\n",
      " |      >>> network.train()\n",
      " |      >>> unperturbed_data = train[((train.obs[\"cell_type\"] == \"CD4T\") & (train.obs[\"condition\"] == \"control\"))]\n",
      " |      >>> pred, delta = network.predict(\n",
      " |      >>>     adata=train,\n",
      " |      >>>     adata_to_predict=unperturbed_data,\n",
      " |      >>>     ctrl_key=\"control\",\n",
      " |      >>>     stim_key=\"stimulated\"\n",
      " |      >>>)\n",
      " |      >>> pred_adata = anndata.AnnData(\n",
      " |      >>>     pred,\n",
      " |      >>>     obs={\"condition\": [\"pred\"] * len(pred)},\n",
      " |      >>>     var={\"var_names\": train.var_names},\n",
      " |      >>>)\n",
      " |      >>> CD4T = train[train.obs[\"cell_type\"] == \"CD4T\"]\n",
      " |      >>> all_adata = CD4T.concatenate(pred_adata)\n",
      " |      >>> network.reg_var_plot(\n",
      " |      >>>     all_adata,\n",
      " |      >>>     axis_keys={\"x\": \"control\", \"y\": \"pred\", \"y1\": \"stimulated\"},\n",
      " |      >>>     gene_list=[\"ISG15\", \"CD3D\"],\n",
      " |      >>>     path_to_save=\"tests/reg_var4.pdf\",\n",
      " |      >>>     show=False\n",
      " |      >>>)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods defined here:\n",
      " |  \n",
      " |  setup_anndata(adata: anndata._core.anndata.AnnData, batch_key: Optional[str] = None, labels_key: Optional[str] = None, **kwargs) from scvi.model.base._base_model.BaseModelMetaClass\n",
      " |      Sets up the :class:`~anndata.AnnData` object for this model.\n",
      " |          A mapping will be created between data fields used by this model to their respective locations in adata.\n",
      " |      \n",
      " |          None of the data in adata are modified. Only adds fields to adata.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |          batch_key\n",
      " |              key in `adata.obs` for batch information. Categories will automatically be converted into integer\n",
      " |              categories and saved to `adata.obs['_scvi_batch']`. If `None`, assigns the same batch to all the data.\n",
      " |          labels_key\n",
      " |              key in `adata.obs` for label information. Categories will automatically be converted into integer\n",
      " |              categories and saved to `adata.obs['_scvi_labels']`. If `None`, assigns the same label to all the data.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      scGen expects the expression data to come from `adata.X`\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  __abstractmethods__ = frozenset()\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from scvi.model.base._vaemixin.VAEMixin:\n",
      " |  \n",
      " |  get_elbo(self, adata: Optional[anndata._core.anndata.AnnData] = None, indices: Optional[Sequence[int]] = None, batch_size: Optional[int] = None) -> float\n",
      " |      Return the ELBO for the data.\n",
      " |      \n",
      " |      The ELBO is a lower bound on the log likelihood of the data used for optimization\n",
      " |      of VAEs. Note, this is not the negative ELBO, higher is better.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      adata\n",
      " |          AnnData object with equivalent structure to initial AnnData. If `None`, defaults to the\n",
      " |          AnnData object used to initialize the model.\n",
      " |      indices\n",
      " |          Indices of cells in adata to use. If `None`, all cells are used.\n",
      " |      batch_size\n",
      " |          Minibatch size for data loading into model. Defaults to `scvi.settings.batch_size`.\n",
      " |  \n",
      " |  get_latent_representation(self, adata: Optional[anndata._core.anndata.AnnData] = None, indices: Optional[Sequence[int]] = None, give_mean: bool = True, mc_samples: int = 5000, batch_size: Optional[int] = None) -> numpy.ndarray\n",
      " |      Return the latent representation for each cell.\n",
      " |      \n",
      " |      This is denoted as :math:`z_n` in our manuscripts.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      adata\n",
      " |          AnnData object with equivalent structure to initial AnnData. If `None`, defaults to the\n",
      " |          AnnData object used to initialize the model.\n",
      " |      indices\n",
      " |          Indices of cells in adata to use. If `None`, all cells are used.\n",
      " |      give_mean\n",
      " |          Give mean of distribution or sample from it.\n",
      " |      mc_samples\n",
      " |          For distributions with no closed-form mean (e.g., `logistic normal`), how many Monte Carlo\n",
      " |          samples to take for computing mean.\n",
      " |      batch_size\n",
      " |          Minibatch size for data loading into model. Defaults to `scvi.settings.batch_size`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      latent_representation : np.ndarray\n",
      " |          Low-dimensional representation for each cell\n",
      " |  \n",
      " |  get_marginal_ll(self, adata: Optional[anndata._core.anndata.AnnData] = None, indices: Optional[Sequence[int]] = None, n_mc_samples: int = 1000, batch_size: Optional[int] = None) -> float\n",
      " |      Return the marginal LL for the data.\n",
      " |      \n",
      " |      The computation here is a biased estimator of the marginal log likelihood of the data.\n",
      " |      Note, this is not the negative log likelihood, higher is better.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      adata\n",
      " |          AnnData object with equivalent structure to initial AnnData. If `None`, defaults to the\n",
      " |          AnnData object used to initialize the model.\n",
      " |      indices\n",
      " |          Indices of cells in adata to use. If `None`, all cells are used.\n",
      " |      n_mc_samples\n",
      " |          Number of Monte Carlo samples to use for marginal LL estimation.\n",
      " |      batch_size\n",
      " |          Minibatch size for data loading into model. Defaults to `scvi.settings.batch_size`.\n",
      " |  \n",
      " |  get_reconstruction_error(self, adata: Optional[anndata._core.anndata.AnnData] = None, indices: Optional[Sequence[int]] = None, batch_size: Optional[int] = None) -> Union[float, Dict[str, float]]\n",
      " |      Return the reconstruction error for the data.\n",
      " |      \n",
      " |      This is typically written as :math:`p(x \\mid z)`, the likelihood term given one posterior sample.\n",
      " |      Note, this is not the negative likelihood, higher is better.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      adata\n",
      " |          AnnData object with equivalent structure to initial AnnData. If `None`, defaults to the\n",
      " |          AnnData object used to initialize the model.\n",
      " |      indices\n",
      " |          Indices of cells in adata to use. If `None`, all cells are used.\n",
      " |      batch_size\n",
      " |          Minibatch size for data loading into model. Defaults to `scvi.settings.batch_size`.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from scvi.model.base._vaemixin.VAEMixin:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from scvi.model.base._training_mixin.UnsupervisedTrainingMixin:\n",
      " |  \n",
      " |  train(self, max_epochs: Optional[int] = None, use_gpu: Union[str, int, bool, NoneType] = None, train_size: float = 0.9, validation_size: Optional[float] = None, batch_size: int = 128, early_stopping: bool = False, plan_kwargs: Optional[dict] = None, **trainer_kwargs)\n",
      " |      Train the model.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      max_epochs\n",
      " |          Number of passes through the dataset. If `None`, defaults to\n",
      " |          `np.min([round((20000 / n_cells) * 400), 400])`\n",
      " |      use_gpu\n",
      " |          Use default GPU if available (if None or True), or index of GPU to use (if int),\n",
      " |          or name of GPU (if str, e.g., `'cuda:0'`), or use CPU (if False).\n",
      " |      train_size\n",
      " |          Size of training set in the range [0.0, 1.0].\n",
      " |      validation_size\n",
      " |          Size of the test set. If `None`, defaults to 1 - `train_size`. If\n",
      " |          `train_size + validation_size < 1`, the remaining cells belong to a test set.\n",
      " |      batch_size\n",
      " |          Minibatch size to use during training.\n",
      " |      early_stopping\n",
      " |          Perform early stopping. Additional arguments can be passed in `**kwargs`.\n",
      " |          See :class:`~scvi.train.Trainer` for further options.\n",
      " |      plan_kwargs\n",
      " |          Keyword args for :class:`~scvi.train.TrainingPlan`. Keyword arguments passed to\n",
      " |          `train()` will overwrite values present in `plan_kwargs`, when appropriate.\n",
      " |      **trainer_kwargs\n",
      " |          Other keyword args for :class:`~scvi.train.Trainer`.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from scvi.model.base._base_model.BaseModelClass:\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return repr(self).\n",
      " |  \n",
      " |  get_anndata_manager(self, adata: anndata._core.anndata.AnnData, required: bool = False) -> Optional[scvi.data._manager.AnnDataManager]\n",
      " |      Retrieves the :class:`~scvi.data.AnnDataManager` for a given AnnData object specific to this model instance.\n",
      " |      \n",
      " |      Requires ``self.id`` has been set. Checks for an :class:`~scvi.data.AnnDataManager`\n",
      " |      specific to this model instance.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      adata\n",
      " |          AnnData object to find manager instance for.\n",
      " |      required\n",
      " |          If True, errors on missing manager. Otherwise, returns None when manager is missing.\n",
      " |  \n",
      " |  get_from_registry(self, adata: anndata._core.anndata.AnnData, registry_key: str) -> numpy.ndarray\n",
      " |      Returns the object in AnnData associated with the key in the data registry.\n",
      " |      \n",
      " |      AnnData object should be registered with the model prior to calling this function\n",
      " |      via the ``self._validate_anndata`` method.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      registry_key\n",
      " |          key of object to get from data registry.\n",
      " |      adata\n",
      " |          AnnData to pull data from.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      The requested data as a NumPy array.\n",
      " |  \n",
      " |  save(self, dir_path: str, prefix: Optional[str] = None, overwrite: bool = False, save_anndata: bool = False, **anndata_write_kwargs)\n",
      " |      Save the state of the model.\n",
      " |      \n",
      " |      Neither the trainer optimizer state nor the trainer history are saved.\n",
      " |      Model files are not expected to be reproducibly saved and loaded across versions\n",
      " |      until we reach version 1.0.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      dir_path\n",
      " |          Path to a directory.\n",
      " |      prefix\n",
      " |          Prefix to prepend to saved file names.\n",
      " |      overwrite\n",
      " |          Overwrite existing data or not. If `False` and directory\n",
      " |          already exists at `dir_path`, error will be raised.\n",
      " |      save_anndata\n",
      " |          If True, also saves the anndata\n",
      " |      anndata_write_kwargs\n",
      " |          Kwargs for :meth:`~anndata.AnnData.write`\n",
      " |  \n",
      " |  to_device(self, device: Union[str, int])\n",
      " |      Move model to device.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      device\n",
      " |          Device to move model to. Options: 'cpu' for CPU, integer GPU index (eg. 0),\n",
      " |          or 'cuda:X' where X is the GPU index (eg. 'cuda:0'). See torch.device for more info.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> adata = scvi.data.synthetic_iid()\n",
      " |      >>> model = scvi.model.SCVI(adata)\n",
      " |      >>> model.to_device('cpu')      # moves model to CPU\n",
      " |      >>> model.to_device('cuda:0')   # moves model to GPU 0\n",
      " |      >>> model.to_device(0)          # also moves model to GPU 0\n",
      " |  \n",
      " |  view_anndata_setup(self, adata: Optional[anndata._core.anndata.AnnData] = None, hide_state_registries: bool = False) -> None\n",
      " |      Print summary of the setup for the initial AnnData or a given AnnData object.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      adata\n",
      " |          AnnData object setup with ``setup_anndata`` or\n",
      " |          :meth:`~scvi.data.AnnDataManager.transfer_fields`.\n",
      " |      hide_state_registries\n",
      " |          If True, prints a shortened summary without details of each state registry.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods inherited from scvi.model.base._base_model.BaseModelClass:\n",
      " |  \n",
      " |  convert_legacy_save(dir_path: str, output_dir_path: str, overwrite: bool = False, prefix: Optional[str] = None) -> None from scvi.model.base._base_model.BaseModelMetaClass\n",
      " |      Converts a legacy saved model (<v0.15.0) to the updated save format.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      dir_path\n",
      " |          Path to directory where legacy model is saved.\n",
      " |      output_dir_path\n",
      " |          Path to save converted save files.\n",
      " |      overwrite\n",
      " |          Overwrite existing data or not. If ``False`` and directory\n",
      " |          already exists at ``output_dir_path``, error will be raised.\n",
      " |      prefix\n",
      " |          Prefix of saved file names.\n",
      " |  \n",
      " |  load(dir_path: str, adata: Optional[anndata._core.anndata.AnnData] = None, use_gpu: Union[str, int, bool, NoneType] = None, prefix: Optional[str] = None, backup_url: Optional[str] = None) from scvi.model.base._base_model.BaseModelMetaClass\n",
      " |      Instantiate a model from the saved output.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      dir_path\n",
      " |          Path to saved outputs.\n",
      " |      adata\n",
      " |          AnnData organized in the same way as data used to train model.\n",
      " |          It is not necessary to run setup_anndata,\n",
      " |          as AnnData is validated against the saved `scvi` setup dictionary.\n",
      " |          If None, will check for and load anndata saved with the model.\n",
      " |      use_gpu\n",
      " |          Load model on default GPU if available (if None or True),\n",
      " |          or index of GPU to use (if int), or name of GPU (if str), or use CPU (if False).\n",
      " |      prefix\n",
      " |          Prefix of saved file names.\n",
      " |      backup_url\n",
      " |          URL to retrieve saved outputs from if not present on disk.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Model with loaded state dictionaries.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> model = ModelClass.load(save_path, adata) # use the name of the model class used to save\n",
      " |      >>> model.get_....\n",
      " |  \n",
      " |  register_manager(adata_manager: scvi.data._manager.AnnDataManager) from scvi.model.base._base_model.BaseModelMetaClass\n",
      " |      Registers an :class:`~scvi.data.AnnDataManager` instance with this model class.\n",
      " |      \n",
      " |      Stores the :class:`~scvi.data.AnnDataManager` reference in a class-specific manager store.\n",
      " |      Intended for use in the ``setup_anndata()`` class method followed up by retrieval of the\n",
      " |      :class:`~scvi.data.AnnDataManager` via the ``_get_most_recent_anndata_manager()`` method in\n",
      " |      the model init method.\n",
      " |      \n",
      " |      Notes\n",
      " |      -----\n",
      " |      Subsequent calls to this method with an :class:`~scvi.data.AnnDataManager` instance referring to the same\n",
      " |      underlying AnnData object will overwrite the reference to previous :class:`~scvi.data.AnnDataManager`.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Static methods inherited from scvi.model.base._base_model.BaseModelClass:\n",
      " |  \n",
      " |  view_setup_args(dir_path: str, prefix: Optional[str] = None) -> None\n",
      " |      Print args used to setup a saved model.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      dir_path\n",
      " |          Path to saved outputs.\n",
      " |      prefix\n",
      " |          Prefix of saved file names.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Readonly properties inherited from scvi.model.base._base_model.BaseModelClass:\n",
      " |  \n",
      " |  adata_manager\n",
      " |      Manager instance associated with self.adata.\n",
      " |  \n",
      " |  device\n",
      " |      The current device that the module's params are on.\n",
      " |  \n",
      " |  history\n",
      " |      Returns computed metrics during training.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from scvi.model.base._base_model.BaseModelClass:\n",
      " |  \n",
      " |  adata\n",
      " |      Data attached to model instance.\n",
      " |  \n",
      " |  is_trained\n",
      " |      Whether the model has been trained.\n",
      " |  \n",
      " |  test_indices\n",
      " |      Observations that are in test set.\n",
      " |  \n",
      " |  train_indices\n",
      " |      Observations that are in train set.\n",
      " |  \n",
      " |  validation_indices\n",
      " |      Observations that are in validation set.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import scgen\n",
    "\n",
    "help(scgen.SCGEN)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bb0e5a1",
   "metadata": {},
   "source": [
    "## scGen\n",
    "\n",
    "Training notebook: `train_scgen.ipynb`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6f9f83f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading scgen_norman19_model1_shuffled\n",
      "\u001b[34mINFO    \u001b[0m File \u001b[35m/home/icb/yuge.ji/projects/feature-attribution-sc/models/scgen_norman19_model1_\u001b[0m\n",
      "         \u001b[35mshuffled/\u001b[0m\u001b[95mmodel.pt\u001b[0m already downloaded                                                \n",
      "loading scgen_norman19_model1_random\n",
      "\u001b[34mINFO    \u001b[0m File \u001b[35m/home/icb/yuge.ji/projects/feature-attribution-sc/models/scgen_norman19_model1_\u001b[0m\n",
      "         \u001b[35mrandom/\u001b[0m\u001b[95mmodel.pt\u001b[0m already downloaded                                                  \n",
      "loading scgen_norman19_model0_shuffled\n",
      "\u001b[34mINFO    \u001b[0m File \u001b[35m/home/icb/yuge.ji/projects/feature-attribution-sc/models/scgen_norman19_model0_\u001b[0m\n",
      "         \u001b[35mshuffled/\u001b[0m\u001b[95mmodel.pt\u001b[0m already downloaded                                                \n",
      "loading scgen_norman19_model4_shuffled\n",
      "\u001b[34mINFO    \u001b[0m File \u001b[35m/home/icb/yuge.ji/projects/feature-attribution-sc/models/scgen_norman19_model4_\u001b[0m\n",
      "         \u001b[35mshuffled/\u001b[0m\u001b[95mmodel.pt\u001b[0m already downloaded                                                \n",
      "loading scgen_norman19_model4_random\n",
      "\u001b[34mINFO    \u001b[0m File \u001b[35m/home/icb/yuge.ji/projects/feature-attribution-sc/models/scgen_norman19_model4_\u001b[0m\n",
      "         \u001b[35mrandom/\u001b[0m\u001b[95mmodel.pt\u001b[0m already downloaded                                                  \n",
      "loading scgen_norman19_model2_random\n",
      "\u001b[34mINFO    \u001b[0m File \u001b[35m/home/icb/yuge.ji/projects/feature-attribution-sc/models/scgen_norman19_model2_\u001b[0m\n",
      "         \u001b[35mrandom/\u001b[0m\u001b[95mmodel.pt\u001b[0m already downloaded                                                  \n",
      "loading scgen_norman19_model2_shuffled\n",
      "\u001b[34mINFO    \u001b[0m File \u001b[35m/home/icb/yuge.ji/projects/feature-attribution-sc/models/scgen_norman19_model2_\u001b[0m\n",
      "         \u001b[35mshuffled/\u001b[0m\u001b[95mmodel.pt\u001b[0m already downloaded                                                \n",
      "loading scgen_norman19_model0_random.pt\n",
      "\u001b[34mINFO    \u001b[0m File \u001b[35m/home/icb/yuge.ji/projects/feature-attribution-sc/models/scgen_norman19_model0_\u001b[0m\n",
      "         \u001b[35mrandom.pt/\u001b[0m\u001b[95mmodel.pt\u001b[0m already downloaded                                               \n",
      "loading scgen_norman19_model3_shuffled\n",
      "\u001b[34mINFO    \u001b[0m File \u001b[35m/home/icb/yuge.ji/projects/feature-attribution-sc/models/scgen_norman19_model3_\u001b[0m\n",
      "         \u001b[35mshuffled/\u001b[0m\u001b[95mmodel.pt\u001b[0m already downloaded                                                \n",
      "loading scgen_norman19_model0_random\n",
      "\u001b[34mINFO    \u001b[0m File \u001b[35m/home/icb/yuge.ji/projects/feature-attribution-sc/models/scgen_norman19_model0_\u001b[0m\n",
      "         \u001b[35mrandom/\u001b[0m\u001b[95mmodel.pt\u001b[0m already downloaded                                                  \n",
      "loading scgen_norman19_model3_random\n",
      "\u001b[34mINFO    \u001b[0m File \u001b[35m/home/icb/yuge.ji/projects/feature-attribution-sc/models/scgen_norman19_model3_\u001b[0m\n",
      "         \u001b[35mrandom/\u001b[0m\u001b[95mmodel.pt\u001b[0m already downloaded                                                  \n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import scgen\n",
    "\n",
    "adata = sc.read(f'{base_path}/datasets/scgen_norman19.h5ad')\n",
    "models = {}\n",
    "for file in os.listdir(f'{base_path}/models'):\n",
    "    if 'scgen' in file:\n",
    "        print('loading', file)\n",
    "        models['_'.join(file.split('_')[1:])] = scgen.SCGEN.load(f'{base_path}/models/{file}', adata=adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "16b9c8f0-7ac4-4353-86bf-b3e574b39be3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['norman19_model1_shuffled', 'norman19_model1_random', 'norman19_model0_shuffled', 'norman19_model4_shuffled', 'norman19_model4_random', 'norman19_model2_random', 'norman19_model2_shuffled', 'norman19_model0_random.pt', 'norman19_model3_shuffled', 'norman19_model0_random', 'norman19_model3_random'])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "26a32869-fa6f-42ab-af6a-2b2715030b1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method _make_data_loader in module scvi.model.base._base_model:\n",
      "\n",
      "_make_data_loader(adata: anndata._core.anndata.AnnData, indices: Optional[Sequence[int]] = None, batch_size: Optional[int] = None, shuffle: bool = False, data_loader_class=None, **data_loader_kwargs) method of scgen._scgen.SCGEN instance\n",
      "    Create a AnnDataLoader object for data iteration.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    adata\n",
      "        AnnData object with equivalent structure to initial AnnData.\n",
      "    indices\n",
      "        Indices of cells in adata to use. If `None`, all cells are used.\n",
      "    batch_size\n",
      "        Minibatch size for data loading into model. Defaults to `scvi.settings.batch_size`.\n",
      "    shuffle\n",
      "        Whether observations are shuffled each iteration though\n",
      "    data_loader_class\n",
      "        Class to use for data loader\n",
      "    data_loader_kwargs\n",
      "        Kwargs to the class-specific data loader class\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(models['norman19_model1_shuffled']._make_data_loader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4f56dd54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCGEN Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">800</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">100</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.2</span>\n",
       "Training status: Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCGEN Model with the following params: \n",
       "n_hidden: \u001b[1;36m800\u001b[0m, n_latent: \u001b[1;36m100\u001b[0m, n_layers: \u001b[1;36m2\u001b[0m, dropout_rate: \u001b[1;36m0.2\u001b[0m\n",
       "Training status: Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCGEN Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">800</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">100</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.2</span>\n",
       "Training status: Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCGEN Model with the following params: \n",
       "n_hidden: \u001b[1;36m800\u001b[0m, n_latent: \u001b[1;36m100\u001b[0m, n_layers: \u001b[1;36m2\u001b[0m, dropout_rate: \u001b[1;36m0.2\u001b[0m\n",
       "Training status: Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCGEN Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">800</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">100</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.2</span>\n",
       "Training status: Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCGEN Model with the following params: \n",
       "n_hidden: \u001b[1;36m800\u001b[0m, n_latent: \u001b[1;36m100\u001b[0m, n_layers: \u001b[1;36m2\u001b[0m, dropout_rate: \u001b[1;36m0.2\u001b[0m\n",
       "Training status: Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCGEN Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">800</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">100</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.2</span>\n",
       "Training status: Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCGEN Model with the following params: \n",
       "n_hidden: \u001b[1;36m800\u001b[0m, n_latent: \u001b[1;36m100\u001b[0m, n_layers: \u001b[1;36m2\u001b[0m, dropout_rate: \u001b[1;36m0.2\u001b[0m\n",
       "Training status: Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCGEN Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">800</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">100</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.2</span>\n",
       "Training status: Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCGEN Model with the following params: \n",
       "n_hidden: \u001b[1;36m800\u001b[0m, n_latent: \u001b[1;36m100\u001b[0m, n_layers: \u001b[1;36m2\u001b[0m, dropout_rate: \u001b[1;36m0.2\u001b[0m\n",
       "Training status: Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCGEN Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">800</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">100</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.2</span>\n",
       "Training status: Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCGEN Model with the following params: \n",
       "n_hidden: \u001b[1;36m800\u001b[0m, n_latent: \u001b[1;36m100\u001b[0m, n_layers: \u001b[1;36m2\u001b[0m, dropout_rate: \u001b[1;36m0.2\u001b[0m\n",
       "Training status: Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCGEN Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">800</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">100</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.2</span>\n",
       "Training status: Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCGEN Model with the following params: \n",
       "n_hidden: \u001b[1;36m800\u001b[0m, n_latent: \u001b[1;36m100\u001b[0m, n_layers: \u001b[1;36m2\u001b[0m, dropout_rate: \u001b[1;36m0.2\u001b[0m\n",
       "Training status: Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCGEN Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">800</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">100</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.2</span>\n",
       "Training status: Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCGEN Model with the following params: \n",
       "n_hidden: \u001b[1;36m800\u001b[0m, n_latent: \u001b[1;36m100\u001b[0m, n_layers: \u001b[1;36m2\u001b[0m, dropout_rate: \u001b[1;36m0.2\u001b[0m\n",
       "Training status: Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCGEN Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">800</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">100</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.2</span>\n",
       "Training status: Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCGEN Model with the following params: \n",
       "n_hidden: \u001b[1;36m800\u001b[0m, n_latent: \u001b[1;36m100\u001b[0m, n_layers: \u001b[1;36m2\u001b[0m, dropout_rate: \u001b[1;36m0.2\u001b[0m\n",
       "Training status: Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCGEN Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">800</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">100</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.2</span>\n",
       "Training status: Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCGEN Model with the following params: \n",
       "n_hidden: \u001b[1;36m800\u001b[0m, n_latent: \u001b[1;36m100\u001b[0m, n_layers: \u001b[1;36m2\u001b[0m, dropout_rate: \u001b[1;36m0.2\u001b[0m\n",
       "Training status: Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">SCGEN Model with the following params: \n",
       "n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">800</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">100</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.2</span>\n",
       "Training status: Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "SCGEN Model with the following params: \n",
       "n_hidden: \u001b[1;36m800\u001b[0m, n_latent: \u001b[1;36m100\u001b[0m, n_layers: \u001b[1;36m2\u001b[0m, dropout_rate: \u001b[1;36m0.2\u001b[0m\n",
       "Training status: Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "{'norman19_model1_shuffled': ,\n",
       " 'norman19_model1_random': ,\n",
       " 'norman19_model0_shuffled': ,\n",
       " 'norman19_model4_shuffled': ,\n",
       " 'norman19_model4_random': ,\n",
       " 'norman19_model2_random': ,\n",
       " 'norman19_model2_shuffled': ,\n",
       " 'norman19_model0_random.pt': ,\n",
       " 'norman19_model3_shuffled': ,\n",
       " 'norman19_model0_random': ,\n",
       " 'norman19_model3_random': }"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "models"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7adde4a4",
   "metadata": {},
   "source": [
    "## scVI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "213e60aa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "571889f6",
   "metadata": {},
   "source": [
    "## scANVI (HLCA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8c395d71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs  n_vars = 587218  2000\n",
       "    obs: 'sample', 'original_celltype_ann', 'study_long', 'study', 'last_author_PI', 'subject_ID', 'subject_ID_as_published', 'pre_or_postnatal', 'age_in_years', 'age_range', 'sex', 'ethnicity', 'mixed_ethnicity', 'smoking_status', 'smoking_history', 'BMI', 'known_lung_disease', 'condition', 'subject_type', 'cause_of_death', 'sample_type', 'anatomical_region_coarse', 'anatomical_region_detailed', 'tissue_dissociation_protocol', 'cells_or_nuclei', 'single_cell_platform', \"3'_or_5'\", 'enrichment', 'sequencing_platform', 'reference_genome_coarse', 'ensembl_release_reference_genome', 'cell_ranger_version', 'disease_status', 'fresh_or_frozen', 'cultured', 'cell_viability_%', 'comments', 'Processing_site', 'dataset', 'anatomical_region_level_1', 'anatomical_region_level_2', 'anatomical_region_level_3', 'anatomical_region_highest_res', 'age', 'ann_level_1', 'ann_level_2', 'ann_level_3', 'ann_level_4', 'ann_level_5', 'ann_highest_res', 'ann_new', 'n_genes', 'total_counts', 'log10_total_counts', 'n_genes_detected', 'mito_frac', 'ribo_frac', 'compl', 'size_factors', 'ann_level_1_clean', 'ann_level_2_clean', 'ann_level_3_clean', 'ann_level_4_clean', 'ann_level_5_clean', 'scanvi_label', 'anatomical_region_ccf_score', 'manual_ann'\n",
       "    var: 'gene_symbols', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection'\n",
       "    uns: 'anatomical_region_level_1_colors', 'ann_level_1_colors', 'ann_level_2_clean_colors', 'ann_level_3_clean_colors', 'hvg', 'manual_ann_colors', 'neighbors', 'pca', 'scanvi_label_colors', 'study_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hlca_path = '/lustre/groups/ml01/workspace/hlca_lisa.sikkema_malte.luecken/HLCA_reproducibility/data/HLCA_core_h5ads/HLCA_v1_integration/HLCA_v1_scANVI_input.h5ad'\n",
    "adata = sc.read(hlca_path)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5be9a67a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mINFO    \u001b[0m File \u001b[35m/home/icb/yuge.ji/projects/HLCA_reproducibility/notebooks/3_atlas_extension/sca\u001b[0m\n",
      "         \u001b[35mnvi_model/\u001b[0m\u001b[95mmodel.pt\u001b[0m already downloaded                                               \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-19 15:20:15.862220: W external/org_tensorflow/tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory\n",
      "2022-09-19 15:20:15.862295: W external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_driver.cc:263] failed call to cuInit: UNKNOWN ERROR (303)\n",
      "/home/icb/till.richter/anaconda3/envs/py37_fa_base/lib/python3.9/site-packages/scvi/data/_utils.py:156: UserWarning: Category 18 in adata.obs['_scvi_labels'] has fewer than 3 cells. Models may not train properly.\n",
      "  warnings.warn(\n",
      "/home/icb/till.richter/anaconda3/envs/py37_fa_base/lib/python3.9/site-packages/scvi/data/_utils.py:156: UserWarning: Category 18 in adata.obs['_scvi_labels'] has fewer than 3 cells. Models may not train properly.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">ScanVI Model with the following params: \n",
       "unlabeled_category: unlabeled, n_hidden: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">128</span>, n_latent: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">30</span>, n_layers: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2</span>, dropout_rate: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0.1</span>, \n",
       "dispersion: gene, gene_likelihood: nb\n",
       "Training status: Trained\n",
       "</pre>\n"
      ],
      "text/plain": [
       "ScanVI Model with the following params: \n",
       "unlabeled_category: unlabeled, n_hidden: \u001b[1;36m128\u001b[0m, n_latent: \u001b[1;36m30\u001b[0m, n_layers: \u001b[1;36m2\u001b[0m, dropout_rate: \u001b[1;36m0.1\u001b[0m, \n",
       "dispersion: gene, gene_likelihood: nb\n",
       "Training status: Trained\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = scvi.model.SCANVI.load('/home/icb/yuge.ji/projects/HLCA_reproducibility/notebooks/3_atlas_extension/scanvi_model/', adata)\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9b7aeedb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AT0': ['SFTPB', 'SCGB3A2', 'SFTA2'],\n",
       " 'AT1': ['CLIC5', 'SPOCK2', 'TIMP3'],\n",
       " 'AT2 proliferating': ['CENPW', 'CDKN3', 'BIRC5'],\n",
       " 'AT2': ['MFSD2A', 'C8orf4', 'C11orf96'],\n",
       " 'Adventitial fibroblasts': ['MFAP5', 'SCARA5', 'PI16'],\n",
       " 'Alveolar Mph CCL3+': ['CCL20', 'FAM89A'],\n",
       " 'Alveolar Mph MT-positive': ['MT1M', 'CCL18', 'MT1E'],\n",
       " 'Alveolar Mph proliferating': ['TYMS', 'CENPW', 'MND1'],\n",
       " 'Alveolar fibroblasts': ['SPINT2', 'LIMCH1', 'FGFR4'],\n",
       " 'Alveolar macrophages': ['CYP27A1', 'MARCO', 'FABP4'],\n",
       " 'B cells': ['MS4A1'],\n",
       " 'Basal resting': ['KRT15', 'KRT17', 'TP63'],\n",
       " 'CD8 T cells': ['CD8A', 'CD8B', 'TRGC2'],\n",
       " 'Classical monocytes': ['S100A12', 'FCN1', 'RNASE2'],\n",
       " 'Club (nasal)': ['SERPINB3', 'TCN1', 'ASRGL1'],\n",
       " 'Club (non-nasal)': ['CYP2F1', 'SCGB3A1', 'BPIFB1'],\n",
       " 'DC1': ['XCR1', 'CLEC9A', 'CLNK'],\n",
       " 'DC2': ['PKIB', 'CLEC10A', 'CD1E'],\n",
       " 'Deuterosomal': ['CDC20B', 'KDELC2', 'E2F7'],\n",
       " 'EC aerocyte capillary': ['SOSTDC1'],\n",
       " 'EC arterial': ['DKK2', 'IGFBP3'],\n",
       " 'EC general capillary': ['IL7R', 'FCN3', 'MT1M'],\n",
       " 'EC venous pulmonary': ['CPE', 'C7', 'MMRN1'],\n",
       " 'EC venous systemic': ['TPD52L1', 'ZNF385D', 'OLFM1'],\n",
       " 'Fibromyocytes': ['FAM83D'],\n",
       " 'Goblet (bronchial)': ['GPX8', 'ALDH1A3', 'CEACAM5'],\n",
       " 'Goblet (nasal)': ['CEACAM5', 'LYPD2', 'PSCA'],\n",
       " 'Goblet (subsegmental)': ['BPIFB1', 'C16orf89', 'NPDC1'],\n",
       " 'Hematopoietic stem cells': ['CRHBP', 'SERPINE2', 'MAP7'],\n",
       " 'Hillock-like': ['DSG3', 'RAB38', 'TNS4'],\n",
       " 'Interstitial Mph perivascular': ['F13A1', 'FOLR2'],\n",
       " 'Ionocyte': ['BSND', 'IGF1', 'CLCNKB'],\n",
       " 'Mast cells': ['MS4A2', 'SLC18A2', 'RGS13'],\n",
       " 'Mesothelium': ['KLK11', 'UPK3B', 'ITLN1'],\n",
       " 'Migratory DCs': ['LAD1', 'CCL19'],\n",
       " 'Multiciliated (nasal)': ['BEST4', 'LYPD2'],\n",
       " 'Multiciliated (non-nasal)': ['RSPH1', 'C20orf85', 'C9orf24'],\n",
       " 'Myofibroblasts': ['LMOD1', 'ATP1B1', 'TYRP1'],\n",
       " 'NK cells': ['GNLY', 'SPON2', 'FCGR3A'],\n",
       " 'Neuroendocrine': ['CELF3', 'SLC6A17', 'CDK5R2'],\n",
       " 'Non-classical monocytes': ['MTSS1', 'LILRA5', 'FCN1'],\n",
       " 'Peribronchial fibroblasts': ['COL15A1', 'CXCL14', 'DIO2'],\n",
       " 'Pericytes': ['LAMC3'],\n",
       " 'Plasma cells': ['TNFRSF17'],\n",
       " 'Plasmacytoid DCs': ['SCT', 'SMPD3', 'LILRA4'],\n",
       " 'SM activated stress response': ['EGR1', 'FOSB', 'ATF3'],\n",
       " 'SMG duct': ['MGST1', 'KRT19', 'KLF5'],\n",
       " 'SMG mucous': ['BPIFB2', 'MUC5B', 'NKX3-1'],\n",
       " 'SMG serous (bronchial)': ['RP11-1143G9.4', 'S100A1', 'PRB3'],\n",
       " 'SMG serous (nasal)': ['STATH', 'ODAM', 'MUC7'],\n",
       " 'Smooth muscle': ['MYH11', 'TINAGL1', 'PLN'],\n",
       " 'Subpleural fibroblasts': ['MARCKSL1', 'APOC1', 'MMP23B'],\n",
       " 'Suprabasal': ['KRT5', 'SERPINB3'],\n",
       " 'T cells proliferating': ['CENPW', 'TK1', 'MKI67'],\n",
       " 'Tuft': ['DAB1'],\n",
       " 'pre-TB secretory': ['SFTPB', 'RNASE1', 'SFTA1P']}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# parse marker gene dictionary from csv for ground truth\n",
    "marker_df = pd.read_csv('/home/icb/yuge.ji/projects/HLCA_reproducibility/notebooks/3_atlas_extension/markergenes.csv', index_col=0)\n",
    "\n",
    "marker_dict = {}\n",
    "for i in range(0, marker_df.shape[1], 3):\n",
    "    ct = marker_df.columns[i].split('_')[0]\n",
    "    l = list(marker_df[\n",
    "        marker_df[f'{ct}_marker_for'].isin([ct, f'{ct} (poss. lowly expressed, non-unique)'])\n",
    "    ][f'{ct}_marker'].values)\n",
    "    # some markers are skipped due to uncertainty in annotation\n",
    "    if len(l) > 0:\n",
    "        marker_dict[ct] = l\n",
    "marker_dict"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (py37_fa_base)",
   "language": "python",
   "name": "py37_fa_base"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
